### PR TITLE
Dark titlebar when using a dark theme on a recent Win10

### PR DESCRIPTION
**Pull request by pmiossec**
## Proposed changes

- Set a dark titlebar when using a dark theme when the user use a supported version of Win10

### Before

The title bar is always white...
image

### After

light theme:

image

dark theme:

image

## Test methodology 

- Manual

## Test environment(s) 

- Git Extensions 33.33.33
- Build 26402f0898e7107f5b8978fa8c8a2416198a864b
- Git 2.27.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by hani77979** [comment]
ي

**Comment by tebeco** [comment]
@pmiossec 

Given the diff shown by github (which include older commits) is it the good base branch (or you might have an older base branch on your fork ?)
image

File: GitUI/GitModuleForm.cs
Comment (RussKie) [inline GitUI/GitModuleForm.cs:93]: Please move it into Interops folder, add a new DwmApi folder, and move this in there.

File: GitUI/GitModuleForm.cs
Comment (RussKie) [inline GitUI/GitModuleForm.cs:74]: Invert

File: GitUI/GitModuleForm.cs
Comment (RussKie) [inline GitUI/GitModuleForm.cs:67]: This may be useful: https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#common-windows-data-types
```suggestion
        private static extern int DwmSetWindowAttribute(IntPtr hwnd, uint attr, ref int attrValue, uint attrSize);
```

**Review by RussKie — COMMENTED** [review]
We should really give credit when copying code 😉 
https://stackoverflow.com/a/62811758/2338036

**Comment by pmiossec** [comment]
> Given the diff shown by github (which include older commits) is it the good base branch (or you might have an older base branch on your fork ?)
> As these commits are part of a PR for colorblind it might be on purpose, just curious though


@tebeco

that's because the branch is based on #8424 (because it introduces change around theme and I wanted to be sure to avoid future conflicts) and so it will be merged after.

**Comment by NikolayXHD** [comment]
@pmiossec fantastic! Few days ago I noticed Telegram messenger app does this, so I thought there must be a way, tried looking it up, failed, and here you are!

**Comment by hani77979** [comment]
F

**Comment by pmiossec** [comment]
>@pmiossec fantastic! Few days ago I noticed Telegram messenger app does this, so I thought there must be a way, tried looking it up, failed, and here you are!

I deserve very few credits.
Credits goes to:
* @halter73 for noticing the possible improvement
* @tebeco for telling it to me and providing the [stackoverflow answer](https://stackoverflow.com/questions/57124243/winforms-dark-title-bar-on-windows-10/62811758#62811758) containing the solution.

I just did a quick dev and the PR. Testing was a little more difficult because my Win10 is too old 😢

**Comment by tebeco** [comment]
I'll take my shoot here since I've spotted this 
* Is that a bot that need to be reported ?
* Is "Code review" open to all Github on `Git Extensions` ?

on this PR:
image
image

on an older PR:
image

**Comment by pmiossec** [comment]
>     * Is that a bot that need to be reported ?

3 or 4 useless comments in 2 or 3 days. So maybe yes...

>     * Is "Code review" open to all Github on `Git Extensions` ?

yes. So, contribute! 😉

**Comment by RussKie** [comment]
Anyone and everyone is welcome to review open (and closed) pull requests.
The final decision to take or reject changes lies with the members of the
core team though.

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (gerhardol) [inline GitUI/Interops/DwmApi/DwmApi.cs:28]:
```suggestion
            var attribute = IsWindows10OrGreater(18985) 
                ? DWMWA_USE_IMMERSIVE_DARK_MODE
                : DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1;
```

**Review by gerhardol — APPROVED** [review]
10.0.19041

**Comment by pmiossec** [comment]
Done

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (drewnoakes) [inline GitUI/Interops/DwmApi/DwmApi.cs:32]: This default is not used anywhere, so remove it.

```suggestion
        private static bool IsWindows10OrGreater(int build) => Environment.OSVersion.Version.Major >= 10 && Environment.OSVersion.Version.Build >= build;
```

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (drewnoakes) [inline GitUI/Interops/DwmApi/DwmApi.cs:19]: Consider either using a named argument:

```suggestion
            if (!IsWindows10OrGreater(build: 17763))
```

Or renaming to `IsWindows10BuildOrGreater(17763)`.

Ditto below.

**Review by drewnoakes — APPROVED** [review]
This is a great improvement to the dark mode. Thanks very much to all involved.

File: GitUI/GitExtensionsForm.cs
Comment (drewnoakes) [inline GitUI/GitExtensionsForm.cs:45]:
```suggestion
                // Warning: This call freezes the CI in AppVeyor, however dark theme is not used on build machines
```

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (drewnoakes) [inline GitUI/Interops/DwmApi/DwmApi.cs:17]: Consider renaming `handle` to `hwnd` to make it clear what kind of handle is expected.

Alternatively accept `Control` or `Window` and dereference the handle directly.

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (RussKie) [inline GitUI/Interops/DwmApi/DwmApi.cs:14]: I had peaked inside the Windows codebase, and these are indeed non-documented private consts
```suggestion
        // Non-documented Windows constants
        private const uint DWMWA_USE_IMMERSIVE_DARK_MODE_BEFORE_20H1 = 19;
```

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (RussKie) [inline GitUI/Interops/DwmApi/DwmApi.cs:11]:
```suggestion
        [DllImport("dwmapi.dll", ExactSpelling = true)]
```

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (RussKie) [inline GitUI/Interops/DwmApi/DwmApi.cs:18]:
```suggestion
        internal static bool UseImmersiveDarkMode(IntPtr hwnd, bool enabled)
        {
            if (hwnd == IntPtr.Zero)
                return false;
```

**Review by RussKie — COMMENTED** [review]
Few more nits

File: GitUI/Interops/DwmApi/DwmApi.cs
Comment (RussKie) [inline GitUI/Interops/DwmApi/DwmApi.cs:23]: Need to check hwnd for validity

Comment (pmiossec) [reply GitUI/Interops/DwmApi/DwmApi.cs:23]: is it enough `hwnd == IntPtr.Zero`? or you see something else?

**Comment by RussKie** [comment]
that's it

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you

**Pull request merged**
